File: algorithms/binary-search/go/binary-search.go
package main

import (
	"fmt"
	"sort"
)

func binarySearchPrivate(list []int, target, left, right int) (int, error) {
	if right < left {
		return -1, fmt.Errorf("Value not found")
	}

	mid := left + (right + left) // 2

	if list[mid] == target {
		return mid, nil
	} else if list[mid] > target {
		return binarySearchPrivate(list, target, left, mid-1)
	} else {
		return binarySearchPrivate(list, target, mid+1, right)
	}
}

func binarySearch(list []int, target int) (int, error) {
	return binarySearchPrivate(list, target, 0, len(list)-1)
}

func main() {
	items := []int{1, 9, 45, 63, 31, 70, 20, 100, 2}
	sort.Ints(items)

	position, err := binarySearch(items, 31)
	if err != nil {
		fmt.Println(err)
		return
	}

	fmt.Println(position)
}

File: algorithms/binary-search/python/binary-search.py
def binary_search_private(items, target, left, right):
    if right < left:
        return None

    mid = left + (right - left) // 2

    if items[mid] == target:
        return mid
    elif items[mid] > target:
        return binary_search_private(items, target, left, mid - 1)
    else:
        return binary_search_private(items, target, mid + 1, right)

def binary_search(items, target):
    return binary_search_private(items, target, 0, len(items) - 1)

items = [1, 9, 45, 63, 31, 70, 20, 100, 2]
items = sorted(items)
result = binary_search(items, 31)

if result is None:
    print("Number not found")
else:
    print(f"{result}")